### ISSUE TITLE
Duplicate values

### ISSUE THREAD

**Issue by nthtran** [paldepind/flyd#40]
Consider this scenario, we want to send values to `s1` on `s2`'s changes:

```js
  const s1 = stream();
  const s2 = stream();

  on(s1, s2);
  on(s1, s2.map( x => x * 2));

  s1.map(x => console.log(x));

  s2(1);
```

This logs:

```
1
1
```

Isn't it supposed to log `1 2` instead?
```
1
2
```

**Comment by nthtran** [comment]
Another example:

```js
const s1 = stream();
const s2 = stream();

s1.map(() => {
  s2(1);
  s2(2);
});

s2.map(x => console.log(x));

s1(null);
```

This logs

```js
2
2
```

instead of

```js
1
2
```

@paldepind any chance of this getting fixed? Really want to use `flyd` but this bug makes it unusable for me atm.

**Comment by paldepind** [comment]
Sorry for the late response. Due to [atomic updates](https://github.com/paldepind/flyd#atomic-updates) the first example should only log a single value.

May I ask why your dependency chain looks like that?

**Comment by nthtran** [comment]
The second example is actually more similar to my use case. I just need to be able to send multiple values down another stream on `flyd.on` or `flyd.map`.

**Comment by paldepind** [comment]
@nthtran Why? Doing so is likely a quite bad idea.

**Comment by nthtran** [comment]
Say for example I have a `response$` stream for server responses. Upon a response I want to trigger 2 actions (e.g. `Action.SetUser` and `Action.SetNotification`) by sending them to an `action$` stream. How would you handle something like this?

**Comment by paldepind** [comment]
You might be able to use [forwardTo](https://github.com/paldepind/flyd-forwardto). Generally pushing values down a stream as in `someStream(someVal)` should be avoided since it is imperative and not reactive. You could also consider a using an action like `Action.SetUserAndNotification`. It is not particularly pretty but it represents that one response triggers an action that does two things.

**Comment by nthtran** [comment]
Yeah `forwardTo` was the first thing I tried and it didn't work either. Something like this:

```js
const s1 = stream();
const s2 = stream();

on(forwardTo(s1, always(1)), s2);
on(forwardTo(s1, always(2)), s2);

s2(1);
```

Still logs `1 1`. Unless I'm missing something?

Sure a solution like `Action.SetUserAndNotification` works. Just a bit surprising that something like the example above seems like it should work fine but doesn't.

**Comment by nthtran** [comment]
Another use case:

```js
const s1 = stream();
const s2 = stream();

s1.map(R.map(s2));
s2.map(x => console.log(x));

s1([0, 1, 2]);
```

With `flyd@0.1.13` which queues only 1 update this logs `2` once. `flyd@0.1.12` logs `2` 3 times.

@paldepind Would you consider this something `flyd` should handle? With the current implementation it seems it'd be difficult to handle both this and atomic updates since new values are set at `s.val` immediately, not treated as discrete events.

**Comment by paldepind** [comment]
As I mentioned before pushing is imperative. In most cases I think it is best avoided.

As you mention yourself this is inferring with atomic updates. In the last example you are pushing three values inside a single atomic update. When you push a value to a stream inside a stream Flyd immediately updates the value of the streams but it enqueues and thus delays the update of the pushed to streams dependencies. Before 0.1.13 the update of a streams dependencies could be enqueued several times and the resulting behavior was quite odd as you noted. Now such an update can only be enqueued once and IMO that is clearly better.

I don't think this can't be handled as well as atomic updates.

**Comment by nthtran** [comment]
I agree that queueing once is better.

I'm still learning more about FRP, how would you achieve what I intended in the last example using a reactive approach?

**Comment by paldepind** [comment]
That depends on the context. Where is the array coming from? How did it get there?

**Comment by nthtran** [comment]
From the server and for each item I want to send a request to another service so I wanted to send each item to another request stream.

**Comment by paldepind** [comment]
If the array originates outside of a stream you could do this: `R.map(someStream, someArray)`.

**Comment by nthtran** [comment]
I currently receive the array from a response stream. Similar to your example here: https://github.com/paldepind/noname-functional-frontend-framework/tree/master/examples/zip-codes.

**Comment by ericgj** [comment]
@nthtran  if you can extract an array of service requests from the original response (in a promise handler) then try what @paldepind suggests on the resolved promise.  Map each onto your service-request stream.  (`R` == Ramda)

Nicely enough, flyd flattens promises automatically, as [you may have seen](https://github.com/paldepind/flyd#using-promises-for-asynchronous-operations).

**Comment by nthtran** [comment]
Thank you for your suggestion @ericgj. Using a promise could work but I prefer a pure stream approach for now. It seems strange for `flyd` to have to rely on promises to handle something streams easily could.

I still don't get how `R.map(someStream, someArray)` is more reactive than `on(R.map(someStream), arrayStream)`. To me the only difference is where the array originates from but that shouldn't matter. Imperative vs reactive aside, It's inconsistent and counterintuitive if you have to do things differently just because something comes from a stream and not outside.

Also, `flyd@0.1.13` breaks `forwardTo`:

```js
const s1 = stream();
s1.map(x => console.log(x));

const s2 = forwardTo(s1, R.identity);

s2(1);
s2(2);
```

This logs `1`. Any subsequent values are ignored.

IMO ideally `flyd` should handle pushing multiple values from within a stream for the sake of consistency, as well as atomic updates. If there are no plans of supporting this then it should be made clear in the README.

@paldepind @ericgj thoughts?

**Comment by nthtran** [comment]
Closed by #45.
